// soundcloud.ts/dist/entities/Me.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.Me = void 0;
var index_1 = require("./index");
var Me = /** @class */ (function () {
    function Me(api) {
        var _this = this;
        this.api = api;
        this.users = new index_1.Users(this.api);
        /**
         * @deprecated
         * Gets your own profile, or your ID if pass in a true param.
         */
        this.get = function (returnID) { return __awaiter(_this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.api.get("/me")];
                    case 1:
                        response = _a.sent();
                        if (returnID)
                            return [2 /*return*/, response.id];
                        return [2 /*return*/, response];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets activities from your homepage.
         */
        this.activities = function () { return __awaiter(_this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.api.get("/me/activities")];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets affiliated activities.
         */
        this.activitiesAffiliated = function () { return __awaiter(_this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.api.get("/me/activities/tracks/affiliated")];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets exclusive activities.
         */
        this.activitiesExclusive = function () { return __awaiter(_this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.api.get("/me/activities/tracks/exclusive")];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets your own activities only.
         */
        this.activitiesOwn = function () { return __awaiter(_this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.api.get("/me/activities/all/own")];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets your app connections, id any.
         */
        this.connections = function () { return __awaiter(_this, void 0, void 0, function () {
            var id, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.get(true)];
                    case 1:
                        id = _a.sent();
                        return [4 /*yield*/, this.api.get("/me/connections")];
                    case 2:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets a connection from its ID.
         */
        this.connection = function (connectionID) { return __awaiter(_this, void 0, void 0, function () {
            var id, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.get(true)];
                    case 1:
                        id = _a.sent();
                        return [4 /*yield*/, this.api.get("/me/connections/" + connectionID)];
                    case 2:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets your tracks.
         */
        this.tracks = function () { return __awaiter(_this, void 0, void 0, function () {
            var id;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.get(true)];
                    case 1:
                        id = _a.sent();
                        return [2 /*return*/, this.users.tracks(id)];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets your comments.
         */
        this.comments = function () { return __awaiter(_this, void 0, void 0, function () {
            var id;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.get(true)];
                    case 1:
                        id = _a.sent();
                        return [2 /*return*/, this.users.comments(id)];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets your favorites.
         */
        this.favorites = function () { return __awaiter(_this, void 0, void 0, function () {
            var id;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.get(true)];
                    case 1:
                        id = _a.sent();
                        return [2 /*return*/, this.users.favorites(id)];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets a favorite.
         */
        this.favorite = function (userResolvable) { return __awaiter(_this, void 0, void 0, function () {
            var id;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.get(true)];
                    case 1:
                        id = _a.sent();
                        return [2 /*return*/, this.users.favorite(id, userResolvable)];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets your followers.
         */
        this.followers = function () { return __awaiter(_this, void 0, void 0, function () {
            var id;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.get(true)];
                    case 1:
                        id = _a.sent();
                        return [2 /*return*/, this.users.followers(id)];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets a follower.
         */
        this.follower = function (userResolvable) { return __awaiter(_this, void 0, void 0, function () {
            var id;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.get(true)];
                    case 1:
                        id = _a.sent();
                        return [2 /*return*/, this.users.follower(id, userResolvable)];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets your followings.
         */
        this.followings = function () { return __awaiter(_this, void 0, void 0, function () {
            var id;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.get(true)];
                    case 1:
                        id = _a.sent();
                        return [2 /*return*/, this.users.followings(id)];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets a following.
         */
        this.following = function (userResolvable) { return __awaiter(_this, void 0, void 0, function () {
            var id;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.get(true)];
                    case 1:
                        id = _a.sent();
                        return [2 /*return*/, this.users.following(id, userResolvable)];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets your playlists.
         */
        this.playlists = function () { return __awaiter(_this, void 0, void 0, function () {
            var id;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.get(true)];
                    case 1:
                        id = _a.sent();
                        return [2 /*return*/, this.users.playlists(id)];
                }
            });
        }); };
        /**
         * @deprecated
         * Gets your social networking profiles.
         */
        this.webProfiles = function () { return __awaiter(_this, void 0, void 0, function () {
            var id;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.get(true)];
                    case 1:
                        id = _a.sent();
                        return [2 /*return*/, this.users.webProfiles(id)];
                }
            });
        }); };
    }
    return Me;
}());
exports.Me = Me;
